Ar/algorithmic strategies (#34)
Ada-Developers-Academy/core-unit-3: algorithmic-strategies/divide-and-conquer.md

> # Divide & Conquer

Reviewer: Nitpick:

```suggestion
# Divide and Conquer
```

> ## Learning Goals
> 
> By the end of this lesson we should be able to:
> 
> - Explain what a divide-and-conquer solution is
> - Write some divide-and-conquer solutions

Reviewer: I made a different straightforward suggestion above, but here's another one...

I think that this topic could use some framing about _why_ students should learn this topic, now. How is it useful? Why should they pay attention?

This is an attempt to hit that goal, and to rewrite this section

```suggestion
## Goals

As developers, we need as many tools to solve problems as possible.

When we understand the _kind_ of solution or algorithm we are using or can use, we can more easily read and understand it.

There are a number of solutions and algorithms classified as using the Divide and Conquer approach. Learning about Divide and Conquer approaches will:

- Help us identify divide and conquer algorithms
- Help us think about, design, and write divide and conquer solutions
```

Author: I like this approach better, especially after you mentioned adding something like this while doing the dynamic programming review. I just updated the casing on divide-and-conquer approach.

> - Write some divide-and-conquer solutions

Author: Added another goal to recognize situations where divide and conquer could be helpful.

## Overview

**Divide and Conquer** is an approach to problem solving that breaks down a large problem into multiple, smaller subproblems. We use the results of those subproblems to solve the original problem.

When we write a divide-and-conquer solution we normally:

1. Break the problem into subproblems of the same type
1. Recursively solve the subproblems
1. Combine the solved subproblems to solve the larger problem

> This name is sometimes also applied to algorithms that only need to solve a single, reduced subproblem without the need to combine any results. In this case, the distinction between a general recursive algorithm, and one that might be called "divide and conquer" is largely based on how much we are able to divide the original problem at each recursive step.

Author: Aw, you noticed that this was a wishy-washy paragraph? Something rubs the wrong way about calling binary search and its ilk "divide and conquer" approaches. I can see why they are: it has to do with the constant rate of the decrease is proportional to the size of the data, rather than some fixed size, but I was having trouble phrasing that concisely, so punted. But you're version works well!

| Divide and Conquer | An algorithm design strategy based on multi-branched recursion. We recursively break down a problem into two or more subproblems of the same or related type, until these become simple enough to be solved directly. The solutions to the subproblems are then combined to give a solution to the original problem. | "If multiple subproblems can be combined to solve a larger problem, we might try to apply a divide-and-conquer solution." |
| Pivot              | An element used in QuickSort to divide the array into two sections, one section less than the pivot and the other greater than the pivot. The choice of the pivot has enormous implications for the efficiency of QuickSort.                                                                                            | "We should be careful about how we pick the pivot so that we avoid worst-case performance for our typical inputs."                                  |

## Example: Binary Search

Binary search can be considered a divide-and-conquer algorithm.

> Starting with a sorted array, we check whether it contains a particular value by comparing the search value with the element in the middle of the array.
> 
> If we find the value we return the position where it was found.
> 
> If we don't find the value, we determine whether it would be in the left or right half of the array by seeing whether it is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
> 
> If at any point, we end up with an empty range, we know the value was not in the array, and we can return a result indicating the value was not found, such as `None`. Other variations of binary search may return the index of where the value _should_ have been, but as a negative value to indicate that it was missing.
> 
> Each recursion divides the array in half and performs the binary search on a smaller subproblem.

Reviewer: What do you think about turning this explanation into a list?

```suggestion
Here's a description of the binary search algorithm:

1. Starting with a sorted array, we check whether it contains a particular value by comparing the search value with the element in the middle of the array.
1. If we find the value we return the position where it was found.
1. If we don't find the value, we determine whether it would be in the left or right half of the array by seeing whether it is smaller or larger than the middle element. Remember, we can make this decision because _we assume the array is sorted!_ Then we perform a binary search on the selected half.
1. If at any point, we end up with an empty range, we know the value was not in the array, and we can return a result indicating the value was not found, such as `None`. Other variations of binary search may return the index of where the value _should_ have been, but as a negative value to indicate that it was missing.
1. Each recursion divides the array in half and performs the binary search on a smaller subproblem.
```

For convenience, here's a screenshot of what this list looks like

<img width="781" alt="Screen Shot 2021-06-09 at 5 09 46 PM" src="https://user-images.githubusercontent.com/30602862/121445025-8074d980-c945-11eb-95e8-7a9dc872d7c4.png">

Author: Nice. I tend not to think in lists, but even just putting the numbers can help anchor the discussion.

> 

Reviewer: I'd love just a clear sentence that calls out the divide and conquer part. There's a similar one for the QuickSort section ("In terms of divide and conquer, we pick a pivot and move smaller elements ...")

I'm not great the grammarz.

```suggestion

In terms of divide and conquer, this algorithm divides the subarray in half with each step. This algorithm will always reduce the subarrays, and thus the amount of remaining data to search through.

```

Author: I see your point, but I think this would be better to go right up at the top when we say binary search can be considered a divide-and-conquer algorithm. Why? this stuff.

_Fig. Tracing through an application of QuickSort in which the final element in a subarray is chosen as the pivot._

In terms of divide and conquer, we pick a pivot and move smaller elements to the left, and larger elements to the right. This leaves us with two smaller subproblems: sorting the elements on the left, and those on the right. We call QuickSort on each side.

If the pivot is relatively well chosen we will divide the list _log n_ levels deep, and shift at most _n_ elements with each level of division, arriving at a _O(n log n)_ runtime.

> **Weakness of QuickSort**: The Pivot

Reviewer: ```suggestion
### Weakness of QuickSort: The Pivot
```

> However, notice we said **if** the pivot is well chosen. If the pivot does not break the list into two relatively equal subarrays it will not arrive at a _O(n log n)_ runtime. Instead it will approach a _O(n<sup>2</sup>)_ runtime.

Reviewer: Transition into this new section. Added a few commas for fun.

```suggestion
Again, QuickSort is _O(n log n)_ **if** the pivot is well-chosen. If the pivot does not break the list into two relatively equal subarrays, it will not arrive at a _O(n log n)_ runtime. Instead, it will approach a _O(n<sup>2</sup>)_ runtime.
```

Author: Thanks for rephrasing the beginning. I didn't really like the exclusive "we" (the instructional staff) as opposed to our more typical inclusive we.

_Fig. Example run of merge sort showing the index calculations used to split and merge the arrays. Note the convention used of the end index being exclusive._

As we can see in the image above, in the first _divide_ step, the original array of size eight gets divided into two subarrays of size four each. This is done by setting the _starting index_ to _0_, the index of the first element in the array, and the _ending index_ is set to one past the last element in the array, using the convention of the ending index being exclusive. The _midway index_ is then computed using the formula:

<div style="margin:auto;width:fit-content;">

_midway index_ = ⌊(_starting index_ + _ending index_) / 2⌋

</div>

Where ⌊⌋ denotes the mathematical _floor_ operation, or integer truncation.

> For the first _divide_ step, the _midway index_ will be _(0+8)/2_ = _4_. In the next _divide_ step, we have two subarrays, one ranging in index from _0_ to _4_ and the other ranging in index from _4_ to _8_. The subarrays are not yet of size one. So, the same action gets repeated to compute the _midway index_. This _divide_ stage continues until the original array of size _n_ is reduced to subarrays of size _1_ each.

Reviewer: Suggestion to break this paragraph down into something that feels like "steps" we can follow. Adjusts some language to express more sequencing.

```suggestion
### Divide

For the first _divide_ step, the _midway index_ will be _(0+8)/2_ = _4_.

In the next _divide_ step, we have two subarrays, one ranging in index from _0_ to _4_ and the other ranging in index from _4_ to _8_.

At this point, the subarrays are not yet of size one. So, the same action gets repeated to compute the _midway index_. This _divide_ stage continues until the original array of size _n_ is reduced to subarrays of size _1_ each.
```

A subarray of size one is trivially sorted, since there can be no elements out of place relative to the single value in the array.

> The _merge_ stage starts by combining two subarrays at a time. While combining the subarrays containing _7_ and _2_ respectively, the value in each is compared, the smaller value, i.e. _2_, is written to the lower index, and the higher value, i.e. _7_, is written to the higher index. Larger merges would continue comparing the values to be merged element by element, taking the smaller until one subarray is "empty," at which point the remainder of the other array can be copied over.

Reviewer: Similar to above, puts this in a section and spaces it out a little more. Minimal grammar adjustments.

```suggestion
### Sort and Merge

The _merge_ stage starts by combining two subarrays at a time.

While combining the subarrays containing _7_ and _2_ respectively, the value in each is compared, the smaller value, i.e. _2_, is written to the lower index, and the higher value, i.e. _7_, is written to the higher index.

Larger merges would continue by comparing the values to be merged element by element, taking the smaller until one subarray is "empty." At that point, the remainder of the other array can be copied over!
```

Author: added that the subarrays are sorted (since this is the sort and merge section)

> An auxiliary array of size _n_ is often used to facilitate the merge steps, after which the values are copied from the auxiliary array back to the original array.

Reviewer: This factoid is important, but breaks the "sequencing" kinda vibe of this section. Suggestion to put it in a callout:

```suggestion
### !callout-info

## Copying and Merging Arrays

How do we write how to merge two arrays? An auxiliary array of size _n_ is often used to facilitate the merge steps, after which the values are copied from the auxiliary array back to the original array.

### !end-callout
```

Author: minor change to stress that we're doing this efficiently (don't need the auxiliary array if we didn't care about that)

The two-way merging continues until there are no more subarrays and the original array is completely sorted.

> Since merge sort always divides the list in half at each divide step, there will be _log n_ levels in the division phase. By keeping data in place, and only adjusting indices to keep track of where the subarrays are located, at each level, there are at most _n_ calculations, for a time complexity of the division phase of _O(n log n)_.

Reviewer: This part transitions out of the algo explanation, and into discussing the complexity.

```suggestion
### Divide and Conquer Complexity in Merge Sort

Since merge sort always divides the list in half at each divide step, there will be _log n_ levels in the division phase. By keeping data in place, and only adjusting indices to keep track of where the subarrays are located, at each level, there are at most _n_ calculations, for a time complexity of the division phase of _O(n log n)_.
```
